linkerd/app/inbound/src/endpoint.rs

```diff
@@ -116,117 +212,119 @@ impl tap::Inspect for Endpoint {
     }
 
     fn route_labels<B>(&self, req: &http::Request<B>) -> Option<Arc<IndexMap<String, String>>> {
-        req.extensions().get::<Route>().map(|r| r.labels().clone())
+        req.extensions()
+            .get::<dst::Route>()
+            .map(|r| r.route.labels().clone())
     }
 
     fn is_outbound<B>(&self, _: &http::Request<B>) -> bool {
         false
     }
 }
 
-impl fmt::Display for Endpoint {
+impl fmt::Display for Target {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         self.addr.fmt(f)
     }
 }
 
-// === impl RecognizeEndpoint ===
-
-impl<A> router::Recognize<http::Request<A>> for RecognizeEndpoint {
-    type Target = Endpoint;
-
-    fn recognize(&self, req: &http::Request<A>) -> Option<Self::Target> {
-        let src = req.extensions().get::<tls::accept::Meta>();
-        debug!("inbound endpoint: src={:?}", src);
-        let addr = src.and_then(|s| s.addrs.target_addr_if_not_local())?;
-
-        let tls_client_id = src
-            .map(|s| s.peer_identity.clone())
-            .unwrap_or_else(|| Conditional::None(tls::ReasonForNoIdentity::Disabled));
-
-        let dst_addr = req
-            .extensions()
-            .get::<DstAddr>()
-            .expect("request extensions should have DstAddr");
-
-        let dst_name = dst_addr.as_ref().name_addr().cloned();
-        let http_settings = dst_addr.http_settings;
-
-        debug!(
-            "inbound endpoint: dst={:?}, proto={:?}",
-            dst_name, http_settings
-        );
-
-        Some(Endpoint {
-            addr,
-            dst_name,
-            http_settings,
-            tls_client_id,
-        })
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::{Endpoint, RecognizeEndpoint};
-    use http;
-    use linkerd2_app_core::{
-        proxy::http::Settings,
-        router::Recognize,
-        transport::{listen, tls},
-        Conditional,
-    };
-    use quickcheck::quickcheck;
-    use std::net;
-
-    fn make_test_endpoint(addr: net::SocketAddr) -> Endpoint {
-        let tls_client_id = TLS_DISABLED;
-        Endpoint {
-            addr,
-            dst_name: None,
-            http_settings: Settings::Http2,
-            tls_client_id,
+impl stack_tracing::GetSpan<()> for Target {
+    fn get_span(&self, _: &()) -> tracing::Span {
+        use tracing::info_span;
+
+        match self.http_settings {
+            http::Settings::Http2 => match self.dst_name.as_ref() {
+                None => info_span!(
+                    "http2",
+                    port = %self.addr.port(),
+                ),
+                Some(name) => info_span!(
+                    "http2",
+                    %name,
+                    port = %self.addr.port(),
+                ),
+            },
```

Comment:
if we update to the latest `tracing` release, we could do
```rust
let span = info_span!("http2", port = %self.addr.port(), name = tracing::field::Empty);
span.record("name", &tracing::field::display(name));
```
which would let us have a single span that may or may not have a `name` field. not sure if the resultant code is actually any clearer, though....